staker/staker_test.gno

```diff
@@ -0,0 +1,412 @@
+	if avg.Cmp(expected) != 0 {
+		t.Errorf("Incorrect average delegated over last. Expected %v, got %v", expected, avg)
+	}
+}
+
+func TestGetCurrentDelegated(t *testing.T) {
+	token := newMockGRC20("Test Token", "TST", 18)
+	s := New(token)
+	delegate1 := std.Address("delegate1")
+	delegate2 := std.Address("delegate2")
+
+	// Set up delegation history for delegate1
+	s.delegationHistory[delegate1] = []DelegationHistory{
+		{amount: u256.NewUint(1000), timestamp: 100},
+		{amount: u256.NewUint(1500), timestamp: 200},
+		{amount: u256.NewUint(2000), timestamp: 300},
+	}
+
+	tests := []struct {
+		name           string
+		delegate       std.Address
+		expectedResult *u256.Uint
+	}{
+		{
+			name:           "Delegate with history",
+			delegate:       delegate1,
+			expectedResult: u256.NewUint(2000),
+		},
+		{
+			name:           "Delegate without history",
+			delegate:       delegate2,
+			expectedResult: u256.NewUint(0),
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			result, err := s.GetCurrentDelegated(tt.delegate)
+			if err != nil {
+				t.Fatalf("Unexpected error: %v", err)
+			}
+			if result.Cmp(tt.expectedResult) != 0 {
+				t.Errorf("Expected %s, got %s", tt.expectedResult.ToString(), result.ToString())
+			}
+		})
+	}
+}
+
+func TestGetDelegatedAt(t *testing.T) {
+	token := newMockGRC20("Test Token", "TST", 18)
+	s := New(token)
+	delegate := std.Address("delegate")
+
+	s.delegationHistory[delegate] = []DelegationHistory{
+		{amount: u256.NewUint(1000), timestamp: 100},
+		{amount: u256.NewUint(1500), timestamp: 200},
+		{amount: u256.NewUint(2000), timestamp: 300},
+	}
+
+	tests := []struct {
+		name           string
+		timestamp      uint64
+		expectedResult *u256.Uint
+	}{
+		{
+			name:           "Before first delegation",
+			timestamp:      50,
+			expectedResult: u256.NewUint(0),
+		},
+		{
+			name:           "At first delegation",
+			timestamp:      100,
+			expectedResult: u256.NewUint(1000),
+		},
+		{
+			name:           "Between first and second delegation",
+			timestamp:      150,
+			expectedResult: u256.NewUint(1000),
+		},
+		{
+			name:           "At second delegation",
+			timestamp:      200,
+			expectedResult: u256.NewUint(1500),
+		},
+		{
+			name:           "At last delegation",
+			timestamp:      300,
+			expectedResult: u256.NewUint(2000),
+		},
+		{
+			name:           "After last delegation",
+			timestamp:      400,
+			expectedResult: u256.NewUint(2000),
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			result, err := s.GetDelegatedAt(delegate, tt.timestamp)
+			if err != nil {
+				t.Fatalf("Unexpected error: %v", err)
+			}
+			if result.Cmp(tt.expectedResult) != 0 {
+				t.Errorf("Expected %s, got %s", tt.expectedResult.ToString(), result.ToString())
+			}
+		})
+	}
+
+	t.Run("Non-existent delegate", func(t *testing.T) {
+		nonExistentDelegate := std.Address("non-existent")
```

Comment:
if you want to generate some random address,

use
```
import "gno.land/p/demo/testutils"

dummyAddr := testutils.TestAddress("dummy") // <- it will generate address g1~~~~ from 'dummy' string
```

Reply:
https://github.com/gnoswap-labs/governance/pull/4/commits/2e380466ac0fe0d7f67ad7cb5d8b045a9d99e8f8